### PR TITLE
qemu_guest_agent: Changing the way to download qemu-ga-win package from virtio-win iso

### PR DESCRIPTION
**Pull request by xiagao**
Qemu-ga-win pkg is not the latest one at present,need to get it from virtio-win iso.
BTW,optimize qemu-ga-instlall/uninstall function.

id: 1514340 

Signed-off-by: Xiaoling Gao <xiagao@redhat.com>

### CODE REVIEW

**Comment by xiagao** [comment]
python ConfigTest.py --testcase=qemu_guest_agent.virtio_serial.gagent_install_uninstall,qemu_guest_agent.virtio_serial.gagent_install --guestname=Win10,RHEL.7.5 --platform=x86_64 --clone=yes

(1/6) smp_8.2048m.repeat1.pre_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.RHEL.7.5.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads: PASS (1952.48 s)
 (2/6) smp_8.2048m.repeat1.pre_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.RHEL.7.5.x86_64.io-github-autotest-qemu.rh_kernel_update: PASS (200.43 s)
 (3/6) smp_8.2048m.repeat1.pre_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Win10.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads: PASS (4727.45 s)
 (4/6) smp_8.2048m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.RHEL.7.5.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.gagent_install: PASS (50.05 s)
 (5/6) smp_8.2048m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Win10.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.gagent_install: PASS (127.24 s)
 (6/6) smp_8.2048m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Win10.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.gagent_install_uninstall: PASS (368.54 s)

File: qemu/tests/qemu_guest_agent.py [hunk 1]
```diff
@@ -117,15 +116,13 @@ def _check_ga_service(self, session, cmd_check_status):
...
-        if args and isinstance(args, tuple):
-            gagent_install_cmd = args[0]
-        else:
-            self.test.error("Missing config 'gagent_install_cmd'")
-
-        if not gagent_install_cmd:
-            self.test.error("Gagent_install_cmd's value is empty.")
-
+    def gagent_install(self, session, vm, gagent_install_cmd):
+        """
```
Comment (kimi1978wy) [inline qemu/tests/qemu_guest_agent.py]: I think "gagent_install_cmd"  can be got in the function, not need to be a param.

Comment (hereischen) [reply qemu/tests/qemu_guest_agent.py]: agree with@kimi1978wy , as the function name is `gagent_install`, I would suggest omit `gagent_install_cmd` for the function, use it directly in the function body.
The same to the `gagent_uninstall`.

File: qemu/tests/qemu_guest_agent.py [hunk 2]
```diff
@@ -905,27 +897,23 @@ class QemuGuestAgentBasicCheckWin(QemuGuestAgentBasicCheck):
...
+        :param session: VM session.
+        :param params: Dictionary with the test parameters
+        :param vm: Virtual machine object.
+        :return qemu_ga_pkg_path: Return the guest agent pkg path.
+        """
+        error_context.context("Get qemu-ga.msi path.", logging.info)
+        qemu_ga_pkg = params["qemu_ga_pkg"]
+        vol_virtio_key = "VolumeName like '%virtio-win%'"
+        vol_virtio = utils_misc.get_win_disk_vol(session, vol_virtio_key)
+        qemu_ga_pkg_path = r"%s:\%s\%s" % (vol_virtio, "guest-agent", qemu_ga_pkg)
```
Comment (hereischen) [inline qemu/tests/qemu_guest_agent.py]: Please use `os.path.join()`
eg: `os.path.join("c:/", "sourcedir")`

Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: as os.path.join is running in linux host,so the result is not suitable for windows guest.

File: qemu/tests/qemu_guest_agent.py (hunk 1 above)
Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: thanks you two,  delete this param.
pls have a check @hereischen @kimi1978wy

File: qemu/tests/qemu_guest_agent.py [hunk 3]
```diff
@@ -937,12 +923,15 @@ def setup(self, test, params, env):
             self.vm = vm
             session = self._get_session(params, self.vm)
 
+            qemu_ga_pkg_path = self.get_qga_pkg_path(session, params, self.vm)
+            self.gagent_install_cmd = params.get("gagent_install_cmd") % qemu_ga_pkg_path
```
Comment (hereischen) [inline qemu/tests/qemu_guest_agent.py]: Is it possible to put `gagent_install_cmd = "start /wait %s /quiet" % self.qemu_ga_pkg_path or qemu_ga_pkg_path` directly into  `gagent_install`?
and same for the uninstall.

**Comment by xiagao** [comment]
(1/3) smp_8.2048m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.RHEL.7.5.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.gagent_install: PASS (49.91 s)
 (2/3) smp_8.2048m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Win10.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.gagent_install: PASS (149.76 s)
 (3/3) smp_8.2048m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Win10.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.gagent_install_uninstall: PASS (446.54 s)

**Comment by hereischen** [comment]
Acked by: Haotong Chen hachen@redhat.com

**Comment by xiagao** [comment]
@luckyh  hello, I have updated this patch. could you help review this patch?

File: qemu/tests/qemu_guest_agent.py [hunk 4]
```diff
@@ -905,27 +895,47 @@ class QemuGuestAgentBasicCheckWin(QemuGuestAgentBasicCheck):
...
+        :param test: kvm test object
+        :param session: VM session.
+        :param params: Dictionary with the test parameters
+        :param vm: Virtual machine object.
+        :return qemu_ga_pkg_path: Return the guest agent pkg path.
+        """
+        error_context.context("Get qemu-ga.msi path where it locate.", logging.info)
+        qemu_ga_pkg = params["qemu_ga_pkg"]
+
+        if params["gagent_download"] == "yes":
```
Comment (luckyh) [inline qemu/tests/qemu_guest_agent.py]: How about make this like:
```python
gagent_src_type = params.get("gagent_src_type", "url")
if gagent_src_type == "url":
    ...
elif gagent_src_type == "virtio-win":
    ...
else:
    ...

File: qemu/tests/qemu_guest_agent.py [hunk 5]
```diff
@@ -937,12 +947,15 @@ def setup(self, test, params, env):
...
+            self.gagent_install_cmd = params.get("gagent_install_cmd") % qemu_ga_pkg_path
+            self.gagent_uninstall_cmd = params.get("gagent_uninstall_cmd") % qemu_ga_pkg_path
+
             if self._check_ga_pkg(session, params.get("gagent_pkg_check_cmd")):
                 logging.info("qemu-ga is already installed.")
             else:
                 logging.info("qemu-ga is not installed.")
-                self.setup_gagent_in_host(session, params, self.vm)
-                self.gagent_install(session, self.vm, *[params.get("gagent_install_cmd")])
+                self.gagent_install(session, self.vm)
```
Comment (luckyh) [inline qemu/tests/qemu_guest_agent.py]: L950 should be performed at this else section as it used to be.

**Comment by xiagao** [comment]
@luckyh  modified. pls have a check.
And test two scenarios.
url :
 (1/1) smp_8.2048m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.gagent_install: PASS (160.20 s)
virtio-win:
(1/1) smp_8.2048m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.gagent_install: PASS (148.14 s)
rhel guest:
 (1/1) smp_8.2048m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.gagent_install: PASS (99.15 s)

**Comment by luckyh** [comment]
Thanks all.

**Pull request merged**